servo/connectors/newrelic.py

```diff
@@ -0,0 +1,340 @@
+
+class NewrelicMetric(Metric):
+    """NewrelicMetric objects describe metrics that can be measure from the Newrelic apm api."""    
+    
+    fetch_name: str
+    """The name of the APM metric containing the values for this Metric
+
+    For details, see the [TODO better Newrelic resource](https://docs.newrelic.com/docs/apis/rest-api-v2/application-examples-v2/average-response-time-examples-v2) documentation.
+    """
+
+    values_selector: str
+    """
+    Values selector  of resultant time series
+    """
+
+    def __init__(self, name: str, unit: Unit, fetch_name: str, values_selector: str) -> None:
+        super().__init__(name=name, unit=unit, fetch_name=fetch_name, values_selector=values_selector, **kwargs)
+
+    def __check__(self) -> Check:
+        return Check(
+            name=f"Check {self.name}",
+            description=f"Run Newrelic get \"{self.fetch_name}: {self.values_selector}\""
+        ) # TODO Checker class
+
+# TODO: what is this used for?
+# class NewrelicTarget():
+#     pass # TODO
+
+class NewrelicConfiguration(BaseConfiguration):
+    """NewrelicConfiguration objects describe how NewrelicConnector objects
+capture measurements from the Newrelic metrics server. 
+    """
+
+    base_url: AnyHttpUrl = DEFAULT_BASE_URL
+    """The base URL for accessing the Newrelic metrics API.
+
+    The URL must point to the root of the Newrelic deployment. Resource paths
+    are computed as necessary for API requests.
+    """
+
+    metrics: List[NewrelicMetric]
+    """The metrics to measure from Newrelic.
+
+    Metrics must include a valid fetch object and values selector.
+    """
+
+    step: Duration = "1m"
+    """The resolution of the metrics to be fetched.
+    
+    The step resolution determines the number of data points captured across a
+    query range.
+    """
+
+    # targets: Optional[List[NewrelicTarget]] TODO: what is this used for?
+    """An optional set of Newrelic target descriptors that are expected to be
+    scraped by the Newrelic instance being queried.
+    """
+
+    @classmethod # TODO
+    def generate(cls, **kwargs) -> "NewrelicConfiguration":
+        """Generates a default configuration for capturing measurements from the
+        Newrelic metrics server.
+
+        Returns:
+            A default configuration for NewrelicConnector objects.
+        """
+        return cls(
+            description="Update the base_url and metrics to match your Newrelic configuration",
+            metrics=[
+                NewrelicMetric(
+                    name="throughput",
+                    unit=Unit.REQUESTS_PER_SECOND,
+                    fetch_name="HttpDispatcher",
+                    values_selector="requests_per_minute"
+                ),
+                NewrelicMetric(
+                    "error_rate", Unit.COUNT, "Errors/all", "error_count"
+                ),
+            ],
+            **kwargs,
+        )
+    
+    @validator("base_url", allow_reuse=True)
```

Comment:
Don't need `allow_reuse`, will clean up on Prometheus (assuming it as cargo culted)